cirq/value/linear_dict.py

```diff
@@ -0,0 +1,173 @@
+
+    LinearDict implements the basic linear algebraic operations of vector
+    addition and scalar multiplication for linear combinations of abstract
+    vectors. Keys represent the vectors, values represent their coefficients.
+    The only requirement on the keys is that they be hashable.
+
+    A consequence of treating keys as opaque is that all relationships between
+    the keys other than equality are ignored. In other words, we treat linear
+    combinations as if all distinct keys are independent.
+
+    For convenience, methods which take a second linear combination accept
+    a standalone key instance as well. The second argument is then treated as
+    a single-term linear combination with the sole coefficient equal to 1.0.
+    """
+    def __init__(self, terms: Mapping[Any, Scalar]) -> None:
+        """Initializes linear combination from a collection of terms.
+
+        Args:
+            terms: Mapping of abstract vectors to coefficients in the linear
+                combination being initialized.
+        """
+        super(LinearDict, self).__init__()
+        self.update(terms)
+
+    def clean(self, *, tolerance: float=1e-9) -> None:
+        """Modifies self by deleting terms with small coefficients."""
+        negligible = [v for v, c in self.items() if abs(c) <= tolerance]
+        for v in negligible:
+            del self[v]
+
+    def copy(self) -> 'LinearDict':
+        return LinearDict(super(LinearDict, self).copy())
+
+    def __getitem__(self, vector: Any) -> Scalar:
+        if vector in self:
+            return super(LinearDict, self).__getitem__(vector)
+        return 0.0
+
+    def __iadd__(self, other: Any) -> 'LinearDict':
+        if not isinstance(other, LinearDict):
+            other = LinearDict({other: 1})
+
+        for vector, coefficient in other.items():
+            self[vector] = self[vector] + coefficient
+        self.clean(tolerance=0)
+        return self
+
+    def __add__(self, other: Any) -> 'LinearDict':
+        result = self.copy()
+        result += other
+        return result
+
+    def __isub__(self, other: Any) -> 'LinearDict':
+        if not isinstance(other, LinearDict):
+            other = LinearDict({other: 1})
+
+        for vector, coefficient in other.items():
+            self[vector] = self[vector] - coefficient
+        self.clean(tolerance=0)
+        return self
+
+    def __sub__(self, other: Any) -> 'LinearDict':
+        result = self.copy()
+        result -= other
+        return result
+
+    def __neg__(self) -> 'LinearDict':
+        return LinearDict({v: -c for v, c in self.items()})
+
+    def __imul__(self, a: Scalar) -> 'LinearDict':
+        for vector in self:
+            self[vector] *= a
+        self.clean(tolerance=0)
+        return self
+
+    def __mul__(self, a: Scalar) -> 'LinearDict':
+        result = self.copy()
+        result *= a
+        return result
+
+    def __rmul__(self, a: Scalar) -> 'LinearDict':
```

Comment:
You need to override `__contains__` to return True iff a key is present with non-zero value.

Reply:
Done.